axum/benches/benches.rs

```diff
@@ -0,0 +1,186 @@
+use axum::{
+    routing::{get, post},
+    Json, Router, Server,
+};
+use hyper::server::conn::AddrIncoming;
+use serde::{Deserialize, Serialize};
+use std::{
+    io::BufRead,
+    process::{Command, Stdio},
+};
+
+fn main() {
+    ensure_rewrk_is_installed();
+
+    benchmark("minimal").run(Router::new);
+
+    benchmark("basic").run(|| Router::new().route("/", get(|| async { "Hello, World!" })));
+
+    benchmark("routing").path("/foo/bar/baz").run(|| {
+        let mut app = Router::new();
+        for a in 0..10 {
+            for b in 0..10 {
+                for c in 0..10 {
+                    app = app.route(&format!("/foo-{}/bar-{}/baz-{}", a, b, c), get(|| async {}));
+                }
+            }
+        }
+        app.route("/foo/bar/baz", get(|| async {}))
+    });
+
+    benchmark("receive-json")
+        .method("post")
+        .headers(&[("content-type", "application/json")])
+        .body(r#"{"n": 123, "s": "hi there", "b": false}"#)
+        .run(|| Router::new().route("/", post(|_: Json<Payload>| async {})));
+
+    benchmark("send-json").run(|| {
+        Router::new().route(
+            "/",
+            get(|| async {
+                Json(Payload {
+                    n: 123,
+                    s: "hi there".to_owned(),
+                    b: false,
+                })
+            }),
+        )
+    });
+}
+
+#[derive(Deserialize, Serialize)]
+struct Payload {
+    n: u32,
+    s: String,
+    b: bool,
+}
+
+fn benchmark(name: &'static str) -> BenchmarkBuilder {
+    BenchmarkBuilder {
+        name,
+        path: None,
+        method: None,
+        headers: None,
+        body: None,
+    }
+}
+
+struct BenchmarkBuilder {
+    name: &'static str,
+    path: Option<&'static str>,
+    method: Option<&'static str>,
+    headers: Option<&'static [(&'static str, &'static str)]>,
+    body: Option<&'static str>,
+}
+
+macro_rules! config_method {
+    ($name:ident, $ty:ty) => {
+        fn $name(mut self, $name: $ty) -> Self {
+            self.$name = Some($name);
+            self
+        }
+    };
+}
+
+impl BenchmarkBuilder {
+    config_method!(path, &'static str);
+    config_method!(method, &'static str);
+    config_method!(headers, &'static [(&'static str, &'static str)]);
+    config_method!(body, &'static str);
+
+    fn run<F>(self, f: F)
+    where
+        F: FnOnce() -> Router,
+    {
+        // support only running some benchmarks with
+        // ```
+        // cargo bench -- routing send-json
+        // ```
+        let args = std::env::args().collect::<Vec<_>>();
+        let names = &args[1..args.len() - 1];
```

Comment:
So wasteful to collect the commandline args once per call of `run` 😭
At least we can do slightly less work w/o complicating the code:

```suggestion
        let names = std::env::args().skip(1).collect::<Vec<_>>();
```